resteasy-client/src/main/java/org/jboss/resteasy/plugins/providers/sse/client/SseEventSourceImpl.java

```diff
@@ -54,6 +58,7 @@
    
    private boolean alwaysReconnect;
 
+   private ClientResponse response;
```

Comment:
Should be volatile

Reply:
Can you please explain why we need volatile ?

Reply:
I don't think this is necessary to be volatile and response is a private field of EventSource.  You have to get the EventSource object to close it.  Moreover, the response in close() won't be null. If it is not connected (not get a response) sucessfully, it will always throw an exception.

Reply:
> Moreover, the response in close() won't be null

Imagine following cases:
**Case1:**
Thread1: user instantiate and `open()` a SseEventSourceImpl
Thread2: this thread starts execution and pause for thread scheduling purpose before executing `response = (ClientResponse)invocation.invoke()` 
( https://github.com/jimma/Resteasy/blob/4c839cdc670b09d461b76be33abd099b1632086d/resteasy-client/src/main/java/org/jboss/resteasy/plugins/providers/sse/client/SseEventSourceImpl.java#L312)
Thread1: user invoke `close()` and a NPE is thrown because `response` has no already be set by Thread2.

**Case2:**
Thread1: user instantiate and `open()` a SseEventSourceImpl
Thread2: this thread start its execution and invokes `response = (ClientResponse)invocation.invoke()` 
( https://github.com/jimma/Resteasy/blob/4c839cdc670b09d461b76be33abd099b1632086d/resteasy-client/src/main/java/org/jboss/resteasy/plugins/providers/sse/client/SseEventSourceImpl.java#L312).
For any reason the connection with the server takes time so that `(ClientResponse)invocation.invoke()` does not return immediately. This thread pause for thread scheduling purpose.
Thread1: user invoke `close()` and a NPE is thrown because `response` has no already be set by Thread2.

Actually many other cases can be found you can even simulate them with debugger.
Since `response` is set and read by two different threads, due to the way thread execution can be scheduled by the system `response` can be null in the reading thread.
And for the same reason even if writing thread set `response` before reading thread tries to read it, there is no guarantee that this thread will read the most up to date `response` value: it can read a null value or an already closed `reponse` in case of reconnect for example. And in this case it will lead to current response never closed.

Reply:
For Case 1, can you show me some code or test case ?  As I said,   response is private and you can't create(write) in thread 1 and close in another thread. 
For Case 2,  volatile can't help fix the NPE if the close EventSource before it is connected successully, right ?  We need to add NPE guard to fix this.

Reply:
>For Case 2, volatile can't help fix the NPE if the close EventSource before it is connected successully, right ? We need to add NPE guard to fix this

Right volatile won't help fixing that, null check will.

>For Case 1, can you show me some code or test case ? As I said, response is private and you can't create(write) in thread 1 and close in another thread.

To simulate the case you can add ` TimeUnit.MINUTES.sleep(2);` before `response = (ClientResponse)invocation.invoke()` in `run` method
( https://github.com/jimma/Resteasy/blob/4c839cdc670b09d461b76be33abd099b1632086d/resteasy-client/src/main/java/org/jboss/resteasy/plugins/providers/sse/client/SseEventSourceImpl.java#L312) to simulate the  thread pause.
And launch the following test:

```
   @Test
   @InSequence(14)
   public void test() throws Exception
   {
      Client client = ClientBuilder.newBuilder().build();
      final AtomicInteger errors = new AtomicInteger(0);
      WebTarget target = client.target(generateURL("/service/server-sent-events"));
      SseEventSourceImpl sourceImpl = (SseEventSourceImpl)SseEventSource.target(target).build();
      sourceImpl.open();
      sourceImpl.close();
   }
```
Thread 1 (test thread): open the  SseEventSource that starts thread 2 in the same time. Thread 1 then waits 30 seconds for the connected latch to be counted down.
Thread 2: pauses for 2 minutes before executing `response=(ClientResponse)invocation.invoke()`
Thread 1: the connected latch timeouts then 'sourceImpl.close();' is invokes then the NPE is thrown.

For both case 1 and case 2 for sure volatile won't fix the issue. The correct fix is to add a null check.
What I was saying about volatile is that: without it we can't say for sure that the closing thread will always read the most up to date value of `response` field.  It is needeed to guarantee that exact `response` set by thread 2 (`run()` method) will always be readable by thread 1 (`close()` method)

Reply:
Hi @asoldano,

To my mind both NPE protection and volatile are required here.